pmlc/dialect/pxa/transforms/cpu_thread.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Intel Corporation
+
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/Pass/Pass.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "pmlc/dialect/pxa/transforms/tile.h"
+
+#include "mlir/Support/DebugStringHelper.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/analysis/uses.h"
+#include "pmlc/dialect/pxa/transforms/autotile.h"
+#include "pmlc/util/logging.h"
+#include "pmlc/util/tags.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+using llvm::DenseMap;
+using llvm::DenseSet;
+using llvm::SmallVector;
+
+// Pick the tiling that:
+// 1) Has at most one index
+// 2) Is as large as possible without going over maxThreads
+struct CostModel {
+  AffineParallelOp op;
+  unsigned maxThreads;
+  CostModel(AffineParallelOp op, unsigned maxThreads)
+      : op(op), maxThreads(maxThreads) {}
+  double operator()(ArrayRef<int64_t> tile, double bestCost) const {
+    int64_t innerSize = 1;
+    for (size_t i = 0; i < tile.size(); i++) {
+      if (tile[i] != 1) {
+        if (tile[i] > maxThreads || innerSize > 1) {
+          return std::numeric_limits<double>::infinity();
+        }
+        innerSize = tile[i];
+      }
+    }
+    return 1.0 / innerSize;
+  }
+};
+
+struct CPUThreadPass : public CPUThreadBase<CPUThreadPass> {
+  CPUThreadPass() = default;
+  explicit CPUThreadPass(unsigned threads) { this->threads = threads; }
```

Comment:
I'm thinking that if `threads` is not specified, the default should be the number of cores on the current machine?

Reply:
How do I do that with the passes.td thing?

Reply:
I.e. I think the base constructor is the one that parses the values from the command line (and sets defaults if they are not there).  So if I overload the default constructor it to get number of threads, won't that break the ability to set it from arguments?

Reply:
perhaps something like the createXSMMStencilPass() function in target/x86/pipeline.cc:98?

Reply:
Right, but XSMMStencilPass doesn't use passes.td to do it's parsing (or registration or anything).  It seems like we would be adding a lot of boilerplate already implemented by passes.td to make this change, considering that pipeline.cc etc will be constructing it in c++ and can pass it the number of threads there, so it's really only pmlc-opt command line usage that we would be fixing.  It's unclear to me if we even want the 'command line' version to have a system dependent default for the number of threads.

Reply:
Realized that the 'default' is just a string that gets pasted into the C++, so it can just as happily be a function.  Implemented.